src/library/Hooks/usePayouts/usePayouts.tsx

```diff
@@ -0,0 +1,163 @@
+/*
+ * Warning: lots of type assertions due to mistyped api library.
+ */
+import { useApi } from 'contexts/Api';
+import { useConnect } from 'contexts/Connect';
+import { useNetworkMetrics } from 'contexts/Network';
+import React, {
+  ReactElement,
+  useContext,
+  useEffect,
+  useMemo,
+  useReducer,
+  useState,
+} from 'react';
+import { round } from 'Utils';
+import asyncSequenceMap from './asyncSequenceMap';
+import calcPayoutForEra from './calcPayoutForEra';
+import fetchDataForEra from './fetchDataForEra';
+import { EraData } from './types';
+
+type ValidatorsData = {
+  [era: number]: EraData;
+};
+
+type ContextType = {
+  validatorsDataCache: ValidatorsData;
+  updateValidatorsDataCache: (patch: ValidatorsData) => void;
+};
+
+const Context = React.createContext<ContextType>({} as ContextType);
+
+/**
+ * Calculates payouts (aka rewards) based on stake data fetched from
+ * the chain. The data is aggressively cached, as it's historical and
+ * immutable. The returned values are in TAZERO (AZERO / 10^12).
+ *
+ * @param lastNEras Fetch last "lastNEras" eras, starting from the first
+ * finished one (the one behind the active one)
+ */
+export default (lastNEras: number) => {
+  const { validatorsDataCache, updateValidatorsDataCache } =
+    useContext(Context);
+  const { activeAccount: validatorId } = useConnect();
+  const { api } = useApi();
+
+  const {
+    metrics: {
+      activeEra: { index: activeEra },
+    },
+  } = useNetworkMetrics();
+  const { selectedErasDataFromCache, erasMissingInCache } = useMemo(
+    () =>
+      [...new Array(Math.min(lastNEras, activeEra || 0))]
+        .map((_, i) => activeEra - i - 1)
+        .reduce<{
+          selectedErasDataFromCache: typeof validatorsDataCache;
+          erasMissingInCache: number[];
+        }>(
+          (acc, eraIndex) => {
+            const eraCache = validatorsDataCache[eraIndex];
+            if (eraCache)
+              return {
+                ...acc,
+                selectedErasDataFromCache: {
+                  ...acc.selectedErasDataFromCache,
+                  [eraIndex]: eraCache,
+                },
+              };
+            return {
+              ...acc,
+              erasMissingInCache: [...acc.erasMissingInCache, eraIndex],
+            };
+          },
+          { selectedErasDataFromCache: {}, erasMissingInCache: [] }
+        ),
+    [activeEra, lastNEras] // Deliberately not putting "validatorsDataCache" in deps since we don't want its change to trigger recalc
+  );
+  const [validatorsData, setValidatorsData] = useState<
+    typeof validatorsDataCache | undefined
+  >(erasMissingInCache.length ? {} : selectedErasDataFromCache);
```

Comment:
This allows us to not display the initial loading spinner if the data is already present in cache. So it's immediately available when you switch pages.